0022-circuit-block.md

```diff
@@ -0,0 +1,152 @@
+
+The overcome these issues, we propose to establish the concept of a block of gates through the introduction of a new `Block` class. The primary goal of this class is to encapsulate an isolated block of `CircuitInstruction`s, treating it as a single unit for tasks like twirling and mitigation. We believe that `Block`s will offer:
+* Full transparency: 
+  * Users will be able to learn how their circuits are broken into blocks, twirled, and mitigated *before* running a job
+  (today, they can only find this out once the job is done).
+* Full flexibility: 
+  * Users will be able to draw blocks around every combination of instructions that they want to be grouped together for the purpose of twirling or mitigation (today, all blocks are effectively layers, i.e., depth-one blocks).
+* Ability to specify a twirling context:
+  * Users will be able to define a twirling group such as "Pauli" and a twirling strategy such as "active-accum" that can potentially be different for each block (today, this can only be done at the circuit level).
+* Ability to assign a noise handle to each block that uniquely specifies its noise model (today, this requires hashing circuits and is inherently unreliable, since circuits are mutable objects).
+
+## `Block`: Definition and use cases
+
+A `Block` represents an isolated block of `CircuitInstruction`s that is treated as a single unit in the context of twirling and mitigation. In more detail, when a circuit contains blocks:
+- Twirling targets the blocks and ignores all the circuit instructions that live outside of the blocks.
+- Noise learning targets the blocks and ignores all the circuit instructions that live outside of the blocks.
+- Gate mitigation (e.g. PEC and PEA) targets the blocks and ignores all the circuit instructions that live outside of the blocks.
+
+We envision two main workflows that users will follow once `Block`s become available, one for regular users and another one for power users:
+1. The workflow for regular users:
+    - Users initialize a `QuantumCircuit` without blocks, as they do today.
+    - They apply all of the desired transiler passes, for example to map the circuit to an ISA circuit for the backend that they wish to use.
+    - They use (a convenience method built around) a new transpiler pass that collects circuit's gates into blocks, with the ability of specifying different collection strategies.
+    - They submit their job.
+2. The workflow for power users:
+    - Users initialize a `QuantumCircuit` adding blocks manually as they wish.
```

Comment:
What does `as they wish` mean specifically? For a regular user, the workflow implies the blocks get added after the circuit is mapped to an ISA circuit. Is the idea that a power user could specify the circuit and blocks at the ISA level or block the circuit before the mapping?

Reply:
Yes, exactly.
It's not hard to directly write a circuit that complies directly with ISA requirements, you just need to use the right gates and have the correct coupling map in front of you. Power users can do this, or also, apply their own passes to the sub-circuits and then inserting them into the bigger circuit as blocks